### ISSUE TITLE
Add awareness for cursor position in selection-related behaviors across desktop platforms

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#57679]
My recollection from working on Gecko macOS text selection many years ago is that desktop platforms have different expected behaviors around how keyboard text selection interacts with existing selections, especially mouse-based. In particular: if you select some text, then shift-left-arrow or shift-right-arrow, what happens? Does it matter if the text was selected from left to right or right to left with the mouse?

We should audit this behavior on each desktop platform in a native text field, then make sure the framework handling matches on each platform.

**Comment by stuartmorgan** [comment]
Related note: @robert-ancell noticed that the base and extent reported by the framework doesn't seem to be different for left-to-right vs right-to-left selection, which may be an issue for getting correct behavior everywhere.

**Comment by justinmc** [comment]
Tl;dr: All native platforms agree on this behavior, but Flutter behaves differently and needs to be updated.

The pivot point is the edge of the selection that remains fixed during the shift/arrow operations before any reversal.  Here is an example of each on Mac:

| Pivot point left | Pivot point right |
| -------------- | ---------------- |
| ![pivot_left](https://user-images.githubusercontent.com/389558/101223853-d5a3fb80-3641-11eb-9bad-83e3127ff8cd.gif) | ![pivot_right](https://user-images.githubusercontent.com/389558/101223858-d9378280-3641-11eb-9f98-e4c390e5cf14.gif) |

For a selection in the middle of the text made with a mouse in the given direction, the behavior is given when holding shift and using the arrow keys.

|    | Flutter | OSX 10.15.7 TextEdit | Windows 10 Home Notepad | Ubuntu 20.04.1 LTS TextEditor |
| - | ------ | --------------------- | ---------------------------- | ------------------------------ |
| Left to right | Pivot point right | Pivot point right | Pivot point right | Pivot point right |
| Right to left | **Pivot point right** | Pivot point left | Pivot point left | Pivot point left |

Is there anything else I'm missing that I should try?  Otherwise it seems that Flutter just needs to adjust its behavior to match all the other platforms (no need to adjust per-platform).

I confirmed that Flutter's [TextSelection](https://master-api.flutter.dev/flutter/services/TextSelection-class.html) has its base set to the lowest point and extent set to the highest, regardless of the direction of the selection.  So indeed, TextSelection does not currently have the information we need to change the pivot point unfortunately.

Here is where we are swapping base and extent to be in ascending order in the codebase:

https://github.com/flutter/flutter/blob/b372c52b2f428c477cb0cdcff9a299ac633b279d/packages/flutter/lib/src/rendering/editable.dart#L1854-L1857

**Comment by justinmc** [comment]
Here's a (WIP) PR to match the native behavior: https://github.com/flutter/flutter/pull/71756

If that doesn't seem to totally break everything, I'll try to move forward with it next week.

**Comment by stuartmorgan** [comment]
Other mouse selection modes like double click followed by keyboard selection adjustment would be good to try. I'm pretty sure there are (were?) at least some edge cases where Windows and macOS disagree about whether to expand or contract selection.

But matching for the simple case would definitely be a good start and could be done first, with edge cases handle later if necessary.

**Comment by justinmc** [comment]
Thanks, I was able to find a bunch of differences!

Here's what happens after double clicking to select:

|    | Flutter | OSX 10.15.7 TextEdit | Windows 10 Home Notepad | Ubuntu 20.04.1 LTS TextEditor |
| - | ------ | --------------------- | ---------------------------- | ------------------------------ |
| double click | Pivot point left | Decided on first arrow key so that selection is expanded | Pivot point left (space selected on double click too) | Pivot point right |

And here's what happens when using word and line modifier keys when moving the selection (like option and command, respectively, on Mac).  The vertical bars indicate selection cursors.  Originally, I'm selecting `word wo|rd| word` with the mouse in the indicated direction.

|    | Flutter | OSX 10.15.7 TextEdit | Windows 10 Home Notepad | Ubuntu 20.04.1 LTS TextEditor |
| - | ------ | --------------------- | ---------------------------- | ------------------------------ |
| Left to right, left + word modifier | word \|wo\|rd word | word \|word\| word | word \|wo\|rd word | word \|wo\|rd word |
| Left to right, right + word modifier | word wo\|rd word\| | word wo\|rd word\| | word wo\|rd \|word | word wo\|rd word\| |
| Left to right, left + line modifier | \|word wo\|rd word | \|word word\| word | n/a | \|word wo\|rd word |
| Left to right, right + line modifier | word wo\|rd word\| | word wo\|rd word\| | n/a | word wo\|rd word\| |
| Right to left, left + word modifier | word \|wo\|rd word | word \|word\| word | word \|word\| word | word \|word\| word |
| Right to left, right + word modifier | word wo\|rd word\| | word wo\|rd word\| | word word\| \|word | word word\| word |
| Right to left, left + line modifier | \|word wo\|rd word | \|word word\| word | n/a | \|word word\| word |
| Right to left, left + line modifier | word wo\|rd word\| | word wo\|rd word\| | n/a | word word\| word\| |

My PR above will not fix most of these differences.  We'll have a lot more platform-specific work to do to get this all 100% correct.

**Comment by justinmc** [comment]
Sorry, this shouldn't have been closed since https://github.com/flutter/flutter/pull/71756 only fixes the first problem mentioned (https://github.com/flutter/flutter/issues/57679#issuecomment-739095036) and not the other problems I found in https://github.com/flutter/flutter/issues/57679#issuecomment-742848741.

**Comment by kf6gpe** [comment]
@justinmc We think this may be done --- can you confirm and close if corrrect? Thanks!

**Comment by justinmc** [comment]
PR https://github.com/flutter/flutter/pull/71756 fixed the main problem, but the issue kind of expanded in scope when I went through all of the inconsistencies on all the different platforms in https://github.com/flutter/flutter/issues/57679#issuecomment-742848741.

I say let's keep this open until I can match all of the behavior in the table in that comment.  Or I can always move that to a separate issue if you think that's better.

**Comment by cbracken** [comment]
@justinmc if we're going to keep this one open rather than open a new bug, it would be useful to update the above table to show which behaviours are working and which still need to be done.

**Comment by justinmc** [comment]
Good call.  I've updated the table in https://github.com/flutter/flutter/issues/57679#issuecomment-742848741 to match the current behavior.  https://github.com/flutter/flutter/pull/75032 will fix some of these and I will update the table again when it's merged.

**Comment by justinmc** [comment]
I've updated the table of behaviors at https://github.com/flutter/flutter/issues/57679#issuecomment-742848741 now that https://github.com/flutter/flutter/pull/75032 has been merged.  It looks like it actually fixed a few of them, but there are still some differences that need to be manually tweaked.

**Comment by justinmc** [comment]
I've created this issue to track Windows inconsistencies: https://github.com/flutter/flutter/issues/78572.

This PR will fix all Mac problems here and will close this issue: https://github.com/flutter/flutter/pull/78573.
